### ISSUE TITLE
Flexibility for yara table configuration and/or extensibility for other security scanners

### ISSUE THREAD

**Issue by mike-myers-tob** [osquery/osquery#6558]
# Blueprint
Desired functionality: a solution to extend osquery table results to include a "scan result" inline with each row of a file-listing-oriented table, showing that the file has been checked by an external (and likely proprietary) scan engine. This would enable the implementation of a security-scan without divulging the scanner's secret sauce (whether signatures or methods). The status quo is represented by the `yara` table, which performs a YARA scan on files, but currently requires the presence of YARA signatures on the local filesystem.

## Driving rationale
Primarily: not be required to manage or divulge local file(s) that contain scanning signatures.
Secondarily: not be limited to the current `yara` table as a scanning engine.

## Summary of last office hours discussion
From the [office hours on the 7th](https://github.com/osquery/foundation/blob/master/docs/office-hours/20200707_meeting_notes.md) the suggestion for adding a generic callback API for external scan-engines was met with a counter proposal to implement this with a SQL table `JOIN`. In the `JOIN` approach, the scan-engine would be an osquery virtual table executing its scan function on each of the enumerated files returned by the first virtual table, and generating the desired table results. This scanner table could be implemented within a (possibly proprietary) osquery extension, as long as it does not require data from evented tables (unavailable to the extension).

Separately, we are looking for discussion feedback on adding other ways for the `yara` table to find its YARA signatures. It was noted that osquery already allows a network connection to be initiated by the Docker and cURL tables, and that "initiate a network connection to retrieve YARA signatures" would be following precedent so we could add that capability. Correct? If we do add the ability to retrieve YARA signatures from a remote host, we'll want to cache them in memory and avoid unnecessary network requests. Now, would it be better to specify a URL to the signature file within the query, or in the osquery config as a new setting?

**Comment by theopolis** [comment]
> The status quo is represented by the `yara` table, which performs a YARA scan on files, but currently requires the presence of YARA signatures on the local filesystem.

Correct. But taking a step back, what if the `yara` table is extended to also read signatures from an osquery configuration? Keep in mind the configuration is cached in the local database, which `root` can read/write. This is to assist osquery if it stops and starts again without a network connection. We want osquery to resume from the last-known-good configuration.

> not be required to manage or divulge local file(s) that contain scanning signatures.

What are you considering 'too difficult for an attacker'? If the files are only readable by `root` there is little difference between storing the signatures in a file and an attacker running `gcore (cat /var/run/osqueryd.pid)`.

> This scanner table could be implemented within a (possibly proprietary) osquery extension

The `JOIN` recommendation is good. But then why not implement `yara_inmemory_scan` as a table within the extension? I need to think more about this but I'm not sure there's a lot of immediate value in abstracting scan concepts, beyond what is already abstracted from a `JOIN` table "API". I can see `scan` being a "front end" table with multiple engines in the "back end" implemented by extensions but this feels like a lot of work to handle the use case.

I think before we can reason about abstractions here we should collect use cases, otherwise I recommend keeping it simple.

> and that "initiate a network connection to retrieve YARA signatures" would be following precedent so we could add that capability

I think this more closely matches the behavior of `carve` vs. `curl`. From a high level, if someone is required to configure the remote host and endpoint via normal osquery configuration options, then this is sane and a recommended way to extend osquery.

> we'll want to cache them in memory and avoid unnecessary network requests

See my above comment about easily dumping user space memory on Linux. What is your threat model exactly?

**Comment by mike-myers-tob** [comment]
Thank you for the prompt initial feedback even if it has taken some time for me to follow-up again here.

We're trying to help clients with what they want and my initial understanding was that the most important goal was to prevent disclosure of YARA rules, when in fact, the primary requirement was to not need to version-control-manage a YARA rules file on all of the hosts since that is expected to be updated frequently and maybe by a separate team that handles threat intelligence. That's not an impossible problem—AV-vendors do it—but it's not built into osquery yet, so it is its own burden.

> what if the `yara` table is extended to also read signatures from an osquery configuration? 

Building on this suggestion, what if the osquery configuration stores only a URL of where to get the signatures, and the `yara` table uses HTTPS to retrieve them whenever they are updated? That is the approach we are going to put up in a PR soon. You may be thinking "well doesn't the osquery agent download its configuration file?" but no, not when used in `osqueryi` mode without an osquery sync server.

Keep in mind the configuration is cached in the local database, which `root` can read/write. This is to assist osquery if it stops and starts again without a network connection. We want osquery to resume from the last-known-good configuration.

> I think this more closely matches the behavior of `carve` vs. `curl`. From a high level, if someone is required to configure the remote host and endpoint via normal osquery configuration options, then this is sane and a recommended way to extend osquery.

> > This scanner table could be implemented within a (possibly proprietary) osquery extension
> 
> why not implement `yara_inmemory_scan` as a table within the extension? 

The intent was to offer a file (or other) enumeration to a scanner (whether YARA or anything else). But it does end up sounding just like a SQL `JOIN` so I think we can disregard the generic scan API suggestion for now.

> I need to think more about this but I'm not sure there's a lot of immediate value in abstracting scan concepts, beyond what is already abstracted from a `JOIN` table "API". I can see `scan` being a "front end" table with multiple engines in the "back end" implemented by extensions but this feels like a lot of work to handle the use case.

Okay! If we really need a generic scan interface we'll do a `JOIN` to some table in an extension that implements it.

> > we'll want to cache them in memory and avoid unnecessary network requests
> 
> See my above comment about easily dumping user space memory on Linux. What is your threat model exactly?

You are absolutely right that user-mode memory can be read as easily as a file owned by root. But the goal of caching here is primarily for performance (use case: `osqueryd` doing a periodic scan using `yara`), and (only secondarily) to avoid a persistent leave-behind file containing the signatures (use case: `osqueryi` being used for a one-shot YARA scan and then exiting).

**Comment by directionless** [comment]
So the intent of this URL approach, as compared to the ad-hoc one in https://github.com/osquery/osquery/pull/6568 to to provide an administrator pre-defined set of signatures?

I think in the config file, this would require some kind of block with a name:rules key:value. Thoughts on that vs selecting with the URL in the query? (Or going all the way and joining the curl table to the yara table)

Regardless, that may have the same data exfiltration that #6568 does, and the same solutions.

**Comment by mike-myers-tob** [comment]
> So the intent of this URL approach, as compared to the ad-hoc one in #6568 to to provide an administrator pre-defined set of signatures?

The new PR to provide signatures by URL is now up, https://github.com/osquery/osquery/pull/6607

Yes, the goal is to allow the administrator to point to a pre-defined set of signatures. But really, the benefit is that it can be an easily maintained and often updated copy of the signatures. Or that you might outsource maintenance of threat intelligence indicators entirely, if this signature set is maintained by someone else.

> I think in the config file, this would require some kind of block with a name:rules key:value. Thoughts on that vs selecting with the URL in the query? (Or going all the way and joining the curl table to the yara table)

The current implementation in #6607 is to specify the URL in the config file _or_ in the query itself. Actually, it can be multiple URLs.
 
> Regardless, that may have the same data exfiltration that #6568 does, and the same solutions.

Yes. To mitigate data exfiltration risk, in the implementation in #6607 :
1. delivering the signatures URL in the query itself requires the new, hidden flag `enable_yara_sigurl` to be set (default: not set)
2. showing the string content that matches the rule will require the new, hidden flag `disable_yara_string_private` (default: not set). The flag's name is a confusing double-negative. I'd prefer `enable_yara_string_results` or something.

In the discussion on #6568 the first mitigation Teddy suggested to remove in favor of just hiding (aka making "private") the strings column _when the signature rule was provided in the query_. Maybe we should do the same thing in #6607

**Comment by theopolis** [comment]
@mike-myers-tob are with finished with this? Can this issue be closed out?

**Comment by mike-myers-tob** [comment]
Yes, closing now thank you everyone for the good feedback.
